book/src/read-only-read-write-accounts.md

```diff
@@ -0,0 +1,95 @@
+# Read-Only/Read-Write Transaction Accounts
+
+This design covers handling of read-only and read-write accounts in the
+[runtime](runtime.md).  Accounts already distinguish themselves as read-only or
+read-write based on the instruction program.  But to identify read-only accounts
+by program id would require the Account to be fetched and loaded from disk.
+This operation is expensive and while it is occurring the runtime would have to
+reject any transactions referencing the same account.
+
+The proposal introduces a `read_only_accounts` field to the Transaction
+structure and removes the `program_ids` dedicated vector for program accounts.
+
+This design doesn't change the runtime transaction processing rules.
+Programs still can't write or spend accounts that they do not own.  But it
+allows the runtime to optimistically take the right lock for each account
+specified in the transaction before loading the accounts from storage.
+
+## Runtime handling
+
+Read-only accounts have the following properties:
+
+* Can be deposited into.  Deposits can be implemented as a simple `atomic_add`.
+
+* Read-only access to userdata.
+
+Instructions that debit or modify the read only account userdata will fail.
+
+## Account Lock Optimizations
+
+The Accounts module that keeps track of current locked accounts in the runtime
+separates read-only accounts from the read-write accounts.  The read-only
+accounts can be cached in memory and shared between all the threads executing
+transactions.
+
+The current runtime cannot predict if an account is read-only or read-write
+when the transaction account references are locked at the entry point.
+Account's referenced by the transaction have not been loaded yet, and loading is
+expensive since it requires a fetch from storage.
+
+An ideal design would cache the read-only accounts while they are referenced by
+any transaction moving through the runtime, and release the cache when the last
+transaction exits the runtime.
+
+## Transaction changes
+
+To enable the possibility of caching accounts only while they are in the
+runtime, the Transaction structure should be changed to contain the following
+additional field:
+
+* `read_only_accounts: u8` - The number of keys from the **end** of the
+Transaction's `account_keys` array that are read-only.
+
+The first account in the `account_keys` array is always the account that pays
+for the transaction fee, therefore it cannot be read-only.  Since accounts can
+only appear once in the transaction's `account_keys` array, an account can only
+be read-only or read-write in a single transaction, not both.
+
+* `program_ids: Vec<Pubkey>` - This vector is removed.  Program keys can be
+placed at the end of the `account_keys` vector with the `read_only_accounts`
+number set to the right number of programs.
+
+## Starvation
```

Comment:
Does this imply that no RW operations can be performed while an account is RO locked for other transactions being processed?  If an account is continuously referenced for RO operations (multiple concurrent transactions) then all transactions that require RW to that account will fail during that time?  Does that also imply that a cached RW request would stall the entire pipeline?

Reply:
@jackcmay This section addresses starvation. Is that not clear?

Reply:
Would like to see more detail here, for example, if a RW operation on a currently RO account is requested the pipeline will stall until the RW operation is performed?  Does this cache RO requests that occur while the RW operation is being processed?

Reply:
@jackcmay 
"While a pending write transaction exists, any additional read lock requests for
that account should fail.  It follows that any other write lock requests will
fail as well."

Reply:
So then yes stall (serialize)

Reply:
@jackcmay not stall locally, the client's responsibility to retransmit.